Feature/testmos

Sorrento110 opened the pull request:
Added test app Causemos to the dojo container stack.

Sorrento110 committed:
Initial commit of testmos feature.

Sorrento110 committed:
Updates after testing the test causemos app.

ccjoel commented on envfile.sample:
```diff
@@ -5,7 +5,8 @@ BIND_PORT=8000
 CAUSEMOS_DEBUG=true
```
Should `CAUSEMOS_DEBUG` be set to false by default in this file as well?

I'm not sure myself: I had not used individual nested project .env files (such as `api/.env`) before. I saw top-level `.env` files were ignored by git in our config, but nested ones aren't.

I do and dmc/ has a different DEBUG flag on, and still points to prod causemos url. Maybe no change for these values are needed, but I'm having a hard time following (maybe not part of this work, but something to track).

What do you think?

ccjoel replied:
An idea I have, which can alleviate this a little bit, is to update the `api/readme` "Running the webapp in development" section to reflect the new changes.

cc @mattprintz in case he can also quickly clarify on these questions. Thanks!

Sorrento110 replied:
we are removing the contents of those sub .env files so they shouldn't be an issue after that change is merged.

ccjoel replied:
Ah- forgot about that very recent change in dev.

ccjoel commented:
Left a question, and will help with testing this soon.

ccjoel commented on api/.env:
```diff
@@ -11,7 +11,7 @@ REDIS_HOST=redis1
+CAUSEMOS_DEBUG=false
 CAUSEMOS_USER=worldmodelers
 CAUSEMOS_PWD=world!
 CAUSEMOS_IND_URL=https://causemos.uncharted.software/api/maas
```
Here CAUSEMOS_IND_URL is not testmos- does that matter? (Will exec into my running api container and check now).

ccjoel replied:
Verified that on my system this created a docker-compose.yaml file with the prod causemos value (not testmos).

ccjoel reviewed (changes requested):
Left a couple of comments. One minor, regarding documentation (api/readme).

I also left a bigger concern. As I understand it, `api/.env` might override the envfile values and still point to causemos prod url for publish. Combining with DEBUG=false would mean that we would publish all to prod during dev, even with testmos running.

I confirmed by rebuilding images, and looking at both the docker-compose.yaml output (for api service) as well as exec'ing into the image and echoing the env var. I did not try to register a dataset since I didn't want to risk publishing a local test to causemos itself.

Sorrento110 commented:
> Left a couple of comments. One minor, regarding documentation (api/readme).
> 
> I also left a bigger concern. As I understand it, `api/.env` might override the envfile values and still point to causemos prod url for publish. Combining with DEBUG=false would mean that we would publish all to prod during dev, even with testmos running.
> 
> I confirmed by rebuilding images, and looking at both the docker-compose.yaml output (for api service) as well as exec'ing into the image and echoing the env var. I did not try to register a dataset since I didn't want to risk publishing a local test to causemos itself.

I should've based this branch off of the branch that removes the contents of the sub .env files, but this branch came first in the development cycle. I am going to make the same change to this branch to prevent these issues from happening.

Sorrento110 committed:
removed contents of sub .env files so env variables are not confused.

ccjoel reviewed (changes requested):
I was not able to get testmos to handle publish requests from the API after testing the latest changes. What I did:
- Pulled dev in, as that contained the envfile.sample for the causemos/sync plugins
- Copied over plugin part of sample to envfile
- Set `CAUSEMOS_DEBUG` env value to false
- docker exec to dojo-api container and verified the above env files were correct.

I restarted all services, rebuilt and restarted the dojo-api, and realized there were some inconsistencies in testmos for the port it was running under (some areas, including envfile.sample, point the service to port 80). I went ahead and replaced the running/exposed/docker/envfile ports to 8012 altogether and tried again.

Even though I took all the actions on the above, registering a new dataset on the UI end to end did not make area requests from the API to testmos. What did I miss?

Sorrento110 committed:
Updated address so that testmos container works across all installations.

Sorrento110 commented:
@ccjoel I was able to get this working on a fresh installation of dojo with the current changes. I had to update the causemos URL in the envfile.sample. With these changes I hope it works for you too.

ccjoel commented:
@Sorrento110 I'm not sure what's going on. I tried it and I finally see the plugin fired, but got this error for `testmos` not being a service in the docker network:

```
2023-05-31 19:39:51,777 INFO:     fastapi - Notifying CauseMos of indicator creation...
2023-05-31 19:39:51,779 DEBUG:    urllib3.connectionpool - Starting new HTTP connection (1): testmos:80
2023-05-31 19:39:51,779 ERROR:    fastapi - Encountered problems communicating with Causemos: HTTPConnectionPool(host='testmos', port=80): Max retries exceeded with url: /api/maas/indicators/post-process (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f9dfa425130>: Failed to establish a new connection: [Errno 111] Connection refused'))
2023-05-31 19:39:51,779 ERROR:    fastapi - HTTPConnectionPool(host='testmos', port=80): Max retries exceeded with url: /api/maas/indicators/post-process (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f9dfa425130>: Failed to establish a new connection: [Errno 111] Connection refused'))
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/connection.py", line 95, in create_connection
    raise err
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/connection.py", line 85, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 714, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 415, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/local/lib/python3.9/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.9/http/client.py", line 950, in send
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 205, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f9dfa425130>: Failed to establish a new connection: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 798, in urlopen
    retries = retries.increment(
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='testmos', port=80): Max retries exceeded with url: /api/maas/indicators/post-process (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f9dfa425130>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./src/plugins/causemos.py", line 77, in notify_causemos
    response = requests.post(
  File "/usr/local/lib/python3.9/site-packages/requests/api.py", line 119, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='testmos', port=80): Max retries exceeded with url: /api/maas/indicators/post-process (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f9dfa425130>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

Sorrento110 committed:
Changed testmos port to 8012 to not conflict on port 80 and for compatibility across machines.

ccjoel committed:
Change api/readme to reflect the possibility of using testmos, as well as change
the mention of api/env to envfile. Modified envfile.sample testmos port to 8012.

ccjoel reviewed (approved):
Worked for both of us on port 8012 🎉 Ready to merge from my perspective.
